doc/design.md

```diff
@@ -0,0 +1,949 @@
+- Data for `Geometry::ComponentInfo` can be mapped to variables in the dataset in a similar manner.
+- `Geometry::Instrument` still exists as part of the pre-**Instrument-2.0** legacy.
+  For the time being it can be supported by simply holding it in a zero-dimensional variable.
+
+##### HistogramData::Histogram
+
+`Dataset` will store histogram data as a single multi-dimensional array, i.e., *not* as a variable holding histograms.
+However, we may still want to operate on a single histogram at a time in some cases.
+A dataset with one-dimensional data variables `Data::Value` and `Data::Variance` alongside a one-dimensional coordinate variable such as `Coord::Tof` is a histogram.
+- Arithmetic operations as well as many of the other basic operations support most of what would be required from a histogram type.
+
+##### DataObjects::EventWorkspace
+
+In contrast to histograms in `DataObjects::Workspace2D`, the event lists stored in `DataObject::EventWorkspace` have varying length.
+Therefore, the events cannot be stored as a single multi-dimensional array.
+- A one-dimensional variable `Data::Events` with `Dim::Spectrum` holds the lists of the events.
+- The lists of events is a type similar to `DataObjects::EventList`.
+  One option is to use `Dataset` for this, that is the dataset would contain many datasets, one for each event list.
+
+##### DataObjects::EventList
+
+A one-dimensional dataset without coordinate variables can be used to represent an event list.
+- `Dim::Event` is used as the only dimension for all variables.
+- `Data::Tof` and `Data::PulseTime` make up the key part of data.
+- Extra variables can be added to support weighted events.
+- `concatenate` provides a merge operation.
+- `sort` provides sorting by time-of-flight or pulse-time.
+- `filter` provides filtering by time-of-flight or pulse-time.
+
+
+### <a name="overview-python-exports"></a>Python exports
+
+All relevant high-level classes and functions are fully exposed via a Python interface.
+- The C++ API is modern and is thus easily mapped to a "pythonic" dataset library.
+- `Dataset` itself is very similar to Python `dict`.
+- Data of variables containing plain old data such as floating point values is exposed as readable and writable `numpy.ndarray`.
+  These are true multi-dimensional arrays, i.e., in contrast to the current `API::MatrixWorkspace::readX()` and `API::MatrixWorkspace::dataX()` it is *not* a separate array for every spectrum.
+
+
+## <a name="design-components"></a>Design components
+
+The following is only an incomplete description.
+For more details and examples the prototype code serves as a reference.
+A good starting point that gives an overview of how `Dataset` may be *used* is given by [`TableWorkspace` example](../test/TableWorkspace_test.cpp), [`Workspace2D` example](../test/Workspace2D_test.cpp), and [`EventWorkspace` example](../test/EventWorkspace_test.cpp).
```

Comment:
Just a suggestion: I know you are linking to the examples but I think adding a very short walk-through one of your examples, a bit like when you showed it to us when Owen & Dan were visiting DMSC, would be useful for the reader to understand, for example how to use the `DatasetView`. Something with a bit more description than the comments in the tests you link to. I can see the `TableWorkspace` is pretty short and easy to read but the `Workspace2D` is a little harder to follow.

Reply:
Good suggestion, I will add something like that.